### PR TITLE
[LICM] Prevent fold and hoist of binary ops with over 2 uses

### PR DESCRIPTION
**Pull request by rj-jesus**
This limits folding and hoisting associative binary ops to cases where the intermediate op has at most two uses.

The more uses the intermediate op has, the more new ops we have to create to potentially reduce the loop's critical path. We keep the limit to two uses to minimise undesirable increases in code size.

This addresses the report in https://github.com/llvm/llvm-project/pull/100377#issuecomment-2252716748.

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/LICM.cpp [hunk 1]
```diff
@@ -2806,7 +2806,8 @@ static bool hoistBOAssociation(Instruction &I, Loop &L,
     return false;
 
   auto *BO0 = dyn_cast<BinaryOperator>(BO->getOperand(0));
-  if (!BO0 || BO0->getOpcode() != Opcode || !BO0->isAssociative())
+  if (!BO0 || BO0->getOpcode() != Opcode || !BO0->isAssociative() ||
+      BO0->getNumUses() > 2)
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/LICM.cpp:2810]:
```suggestion
      BO0->hasNUsesOrMore(3))
```
Early exit use list scan.

**Review by nikic — COMMENTED** [review]
This looks reasonable to me, but @dtcxzyw should confirm it fixes all the regressions.

File: llvm/lib/Transforms/Scalar/LICM.cpp (hunk 1 above)
Comment (rj-jesus) [reply llvm/lib/Transforms/Scalar/LICM.cpp]: Thanks, done.

**Review by dtcxzyw — APPROVED** [review]
All regressions have been fixed :) Thank you!

**Comment by rj-jesus** [comment]
Perfect, thank you for checking! :)

**Pull request merged**